src/core/control/tools/StrokeHandler.cpp

```diff
@@ -101,61 +66,47 @@ auto StrokeHandler::onMotionNotifyEvent(const PositionInputData& pos, double zoo
         if (distance < PIXEL_MOTION_THRESHOLD) {  //(!validMotion(point, endPoint)) {
             if (pointCount == 1 && this->hasPressure && endPoint.z < point.z) {
                 // Record the possible increase in pressure for the first point
-                // Nb: at this point, neither point.z nor endPoint.z has been multiplied by this->stroke->getWidth()
                 this->stroke->setLastPressure(point.z);
-                this->firstPointPressureChange = true;
-
-                double width = this->stroke->getWidth() * point.z;
-                if (mask) {
-                    this->paintDot(mask->cr, endPoint.x, endPoint.y, width);
-                }
-                // Trigger a call to `draw`. If mask == nullopt, the `paintDot` is called in `draw`
-                this->pageView->repaintRect(endPoint.x - 0.5 * width, endPoint.y - 0.5 * width, width, width);
+                this->viewPool->dispatch(xoj::view::StrokeToolView::THICKEN_FIRST_POINT_REQUEST, point.z);
             }
             return;
         }
         if (this->hasPressure) {
             /**
              * Both device and tool are pressure sensitive
              */
-            if (this->firstPointPressureChange) {
-                // Avoid shrinking if we recorded a higher pressure event at the beginning of the stroke
-                this->firstPointPressureChange = false;
-                stroke->setLastPressure(std::max(endPoint.z, point.z) * stroke->getWidth());
-            } else {
-                if (const double widthDelta = (point.z - endPoint.z) * stroke->getWidth();
-                    - widthDelta > MAX_WIDTH_VARIATION || widthDelta > MAX_WIDTH_VARIATION) {
-                    /**
-                     * If the width variation is to big, decompose into shorter segments.
-                     * Those segments can not be shorter than PIXEL_MOTION_THRESHOLD
-                     */
-                    double nbSteps = std::min(std::ceil(std::abs(widthDelta) / MAX_WIDTH_VARIATION),
-                                              std::floor(distance / PIXEL_MOTION_THRESHOLD));
-                    double stepLength = 1.0 / nbSteps;
-                    Point increment((point.x - endPoint.x) * stepLength, (point.y - endPoint.y) * stepLength,
-                                    widthDelta * stepLength);
-                    endPoint.z *= stroke->getWidth();
+            if (const double widthDelta = point.z - endPoint.z;
+                - widthDelta > MAX_WIDTH_VARIATION || widthDelta > MAX_WIDTH_VARIATION) {
+                /**
+                 * If the width variation is to big, decompose into shorter segments.
+                 * Those segments can not be shorter than PIXEL_MOTION_THRESHOLD
+                 */
+                double nbSteps = std::min(std::ceil(std::abs(widthDelta) / MAX_WIDTH_VARIATION),
+                                          std::floor(distance / PIXEL_MOTION_THRESHOLD));
```

Comment:
Should add a comment explaining that `nbSteps >= 2.0` because `distance >= PIXEL_MOTION_THRESHOLD` and `abs(widthDelta) > MAX_WIDTH_VARIATION`. This fact was not very clear to me on the first reading.

Reply:
No, nbSteps could still be 1 (e.g. if distance == 1.1 * PIXEL_MOTION_THRESHOLD), but this is not a problem: the last (and only) point is added by the call to `drawSegmentTo` at the end of the function.

Reply:
Added a comment

Reply:
Right, since the operation is `min`, your example would make `nbSteps == 1`.